apps/labs/posts/unlocking-c-level-performance-in-df-apply.md

```diff
@@ -0,0 +1,466 @@
+method that does what you want.
+
+In pandas 2.2.0, though, a new engine (`engine="numba"`) option will be added to `DataFrame.apply`, opening up the possibility for fast and parallelizable apply in pandas. It works by using numba, a JIT (just-in-time) compiler that translates your Python/numpy functions into fast machine code when their called, providing a several times speedup compared to the original python engine.
+
+So, can I just slap on `engine="numba"` to `DataFrame.apply` and enjoy a free many-times speedup?
+
+```diff
+- df.apply(f)
++ df.apply(f, engine="numba")
+```
+
+Well, not quite. Because of the way that numba/the numba engine inside pandas works, there are several caveats and gotchas to keep in mind to obtain this speedup.
+
+TLDR:
+Use apply only when you have to (e.g. there is no corresponding method or chain of pandas methods that do that you want).
+
+## How the numba engine works
+
+Currently, in pandas, there are currently two implementations of the numba engine inside of apply, one for `raw=True`, and one for `raw=False` (the default).
+
+If you didn't know already, passing in `raw=True` to `DataFrame.apply` tells pandas to pass in numpy arrays, not pandas Series to the functions you pass into apply. Because, numba itself supports numpy arrays natively, in general, every function runnable with the `Python` engine, should also be runnable with the `numba` engine, as long as you stick to the [supported numpy features](https://numba.readthedocs.io/en/latest/reference/numpysupported.html) inside numba.
+
+When `raw=False` is passed in or no `raw` parameter is passed in, we must "mock" the pandas `Series` object that would normally get passsed in, since numba doesn't natively support pandas DataFrames. This means that while most functions passed into apply with work with the numba engine out of the box with `raw=False`, some functions may require modifications, as the "mocked" DataFrame does not implement the entire pandas API.
+
+### Supported features
+
+To summarize the existing support as of pandas 2.2.0, the `numba` engine in df.apply supports:
+
+- Numeric numpy arrays
+- Series methods \*
+  - In general, all methods that have a numpy equivalent (e.g. mean) are supported
+  - Other more complex methods are generally unsupported.
+- Indexing \*
+  - Indexing with a scalar is supported with the numba engine, but
+    indexing with a slice object or via fancy indexing is not supported.
+- Basic binary operations
+  - e.g. addition, subtraction, multiplication, and division
+- Parallel apply
+  - This is only supported when `raw=True`
+  - Like other functions that support numba, you can pass in `parallel: True` inside the `engine_kwargs` dictionary argument to use apply in parallel.
+
+\*: This indicates partial support
+
+Because the numba engine is still very new and experimental, support does not include for now:
+
+- Support for Apache Arrow arrays
```

Comment:
The bullet lists are visually not great when used too heavily. How about making this a regular paragraph instead?

Reply:
Updated.